msal/managed_identity.py

```diff
@@ -0,0 +1,592 @@
+                    data={},
+                ))
+            if (result and "error" not in result) or (not access_token_from_cache):
+                return result
+        except:  # The exact HTTP exception is transportation-layer dependent
+            # Typically network error. Potential AAD outage?
+            if not access_token_from_cache:  # It means there is no fall back option
+                raise  # We choose to bubble up the exception
+        return access_token_from_cache
+
+
+def _scope_to_resource(scope):  # This is an experimental reasonable-effort approach
+    u = urlparse(scope)
+    if u.scheme:
+        return "{}://{}".format(u.scheme, u.netloc)
+    return scope  # There is no much else we can do here
+
+
+APP_SERVICE = object()
+AZURE_ARC = object()
+CLOUD_SHELL = object()  # In MSAL Python, token acquisition was done by
+    # PublicClientApplication(...).acquire_token_interactive(..., prompt="none")
+MACHINE_LEARNING = object()
+SERVICE_FABRIC = object()
+DEFAULT_TO_VM = object()  # Unknown environment; default to VM; you may want to probe
+def get_managed_identity_source():
+    """Detect the current environment and return the likely identity source.
+
+    When this function returns ``CLOUD_SHELL``, you should use
+    :func:`msal.PublicClientApplication.acquire_token_interactive` with ``prompt="none"``
+    to obtain a token.
+    """
+    if ("IDENTITY_ENDPOINT" in os.environ and "IDENTITY_HEADER" in os.environ
+            and "IDENTITY_SERVER_THUMBPRINT" in os.environ
+    ):
+        return SERVICE_FABRIC
+    if "IDENTITY_ENDPOINT" in os.environ and "IDENTITY_HEADER" in os.environ:
+        return APP_SERVICE
+    if "MSI_ENDPOINT" in os.environ and "MSI_SECRET" in os.environ:
+        return MACHINE_LEARNING
+    if "IDENTITY_ENDPOINT" in os.environ and "IMDS_ENDPOINT" in os.environ:
+        return AZURE_ARC
+    if _is_running_in_cloud_shell():
+        return CLOUD_SHELL
+    return DEFAULT_TO_VM
+
+
+def _obtain_token(http_client, managed_identity, resource):
+    # A unified low-level API that talks to different Managed Identity
+    if ("IDENTITY_ENDPOINT" in os.environ and "IDENTITY_HEADER" in os.environ
+            and "IDENTITY_SERVER_THUMBPRINT" in os.environ
+    ):
+        if managed_identity:
+            logger.debug(
+                "Ignoring managed_identity parameter. "
+                "Managed Identity in Service Fabric is configured in the cluster, "
+                "not during runtime. See also "
+                "https://learn.microsoft.com/en-us/azure/service-fabric/configure-existing-cluster-enable-managed-identity-token-service")
+        return _obtain_token_on_service_fabric(
+            http_client,
+            os.environ["IDENTITY_ENDPOINT"],
+            os.environ["IDENTITY_HEADER"],
+            os.environ["IDENTITY_SERVER_THUMBPRINT"],
+            resource,
+        )
+    if "IDENTITY_ENDPOINT" in os.environ and "IDENTITY_HEADER" in os.environ:
```

Comment:
The detection conditions used here and in `get_managed_identity_source` are very fragile. There is no document claiming these environment variables can uniquely identify a managed identity type.

For example, if a managed identity not on App Service also exposes `IDENTITY_ENDPOINT` and `IDENTITY_HEADER`, the name `_obtain_token_on_app_service` is no longer accurate and may even result in wrong logic. If Azure Arc exposes `IDENTITY_HEADER` one day, it'll trick MSAL into the code for App Service.

Reply:
This is a managed identity design issue which we will address with MSIv2, which is currently being designed. The logic here is what Azure SDK have everywhere. 

In particular to your comment, you cannot install Azure Arc on any Azure host. Azure Arc is about importing your own VM or another cloud host into Azure.

Reply:
> In particular to your comment, you cannot install Azure Arc on any Azure host. Azure Arc is about importing your own VM or another cloud host into Azure.

I understand. This is merely an example showing the fragility of using available environment variables to detect the platform.

Reply:
This was discussed in an offline conversation. We agreed that it is not a blocking issue to the release.